Modules/md5module.c

```diff
@@ -318,21 +318,28 @@ md5_done(struct md5_state *md5, unsigned char *out)
  * ------------------------------------------------------------------------
  */
 
-static PyTypeObject MD5type;
+typedef struct {
+    PyTypeObject* md5_type;
+} MD5State;
 
+static inline MD5State* md5_get_state(PyObject *module) {
+    void *state = PyModule_GetState(module);
+    assert(state != NULL);
+    return (MD5State *)state;
+}
 
 static MD5object *
-newMD5object(void)
+newMD5object(MD5State * st)
 {
-    return (MD5object *)PyObject_New(MD5object, &MD5type);
+    return (MD5object *)PyObject_New(MD5object, st->md5_type);
 }
 
-
 /* Internal methods for a hash object */
 
 static void
 MD5_dealloc(PyObject *ptr)
 {
+    Py_DECREF(Py_TYPE(ptr));
```

Comment:
Please DECREF the type at the end. See Modules/_abc.c dealloc function for example.

Same remark for other dealloc functions.